substrate/frame/contracts/src/transient_storage.rs

```diff
@@ -0,0 +1,693 @@
+		self.0.get(key).cloned()
+	}
+
+	/// Write the storage entry.
+	pub fn write(&mut self, key: &Vec<u8>, value: Option<Vec<u8>>) -> Option<Vec<u8>> {
+		if let Some(value) = value {
+			// Insert storage entry.
+			self.0.insert(key.clone(), value)
+		} else {
+			// Remove storage entry.
+			self.0.remove(key)
+		}
+	}
+}
+
+/// Transient storage behaves almost identically to regular storage but is discarded after each
+/// transaction. It consists of a `BTreeMap` for the current state, a journal of all changes, and a
+/// list of checkpoints. On entry to the `start_transaction` function, a marker (checkpoint) is
+/// added to the list. New values are written to the current state, and the previous value is
+/// recorded in the journal (`write`). When the `commit_transaction` function is called, the marker
+/// to the journal index (checkpoint) of when that call was entered is discarded.
+/// On `rollback_transaction`, all entries are reverted up to the last checkpoint.
+pub struct TransientStorage<T: Config> {
+	// The storage and journal size is limited by the storage meter.
+	storage: Storage,
+	journal: Journal,
+	// The size of the StorageMeter is limited by the stack depth.
+	meter: StorageMeter<T>,
+	// The size of the checkpoints is limited by the stack depth.
+	checkpoints: Vec<usize>,
+}
+
+impl<T: Config> TransientStorage<T> {
+	/// Create new transient storage with the supplied memory limit.
+	pub fn new(memory_limit: u32) -> Self {
+		TransientStorage {
+			storage: Default::default(),
+			journal: Journal::new(),
+			checkpoints: Default::default(),
+			meter: StorageMeter::new(memory_limit),
+		}
+	}
+
+	/// Read the storage value. If the entry does not exist, `None` is returned.
+	pub fn read(&self, account: &AccountIdOf<T>, key: &Key<T>) -> Option<Vec<u8>> {
+		self.storage.read(&Self::storage_key(&account.encode(), &key.hash()))
+	}
+
+	/// Write a value to storage.
+	///
+	/// If the `value` is `None`, then the entry is removed. If `take` is true,
+	/// a [`WriteOutcome::Taken`] is returned instead of a [`WriteOutcome::Overwritten`].
+	/// If the entry did not exist, [`WriteOutcome::New`] is returned.
+	pub fn write(
+		&mut self,
+		account: &AccountIdOf<T>,
+		key: &Key<T>,
+		value: Option<Vec<u8>>,
+		take: bool,
+	) -> Result<WriteOutcome, DispatchError> {
+		let prev_value = self.read(account, key);
+		// Skip if the same value is being set.
+		if prev_value != value {
+			let key = Self::storage_key(&account.encode(), &key.hash());
+
+			// Calculate the allocation size.
+			if let Some(value) = &value {
+				// Charge the key, value and journal entry.
+				// If a new value is written, a new journal entry is created. The previous value is
+				// moved to the journal along with its key, and the new value is written to
+				// storage.
+				let key_len = key.capacity();
+				let mut amount = value
+					.capacity()
+					.saturating_add(key_len)
+					.saturating_add(mem::size_of::<JournalEntry>());
```

Comment:
Why are you using `capacity` and not `len`?

Reply:
I used it to track reserved memory to ensure safety, but assuming the lazy memory allocation model in WASM —where physical memory is not committed until it is actually used—using len might be sufficient.

Reply:
We are diving deep into operating system specifics here to know when this memory is actually faulted in. I guess we leave it with capacity just to be sure.